spring-cloud-gcp-core/src/main/java/org/springframework/cloud/gcp/core/GcpEnvironment.java

```diff
@@ -0,0 +1,76 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.core;
+
+import java.util.function.BooleanSupplier;
+
+import com.google.cloud.MetadataConfig;
+import com.google.cloud.ServiceOptions;
+
+/**
+ * Enumeration of valid individual GCP environments.
+ *
+ * <p>Also includes meta-members for convenient evaluation of environment groups, such as {@code ANY_APP_ENGINE}.
+ *
+ * @author Elena Felder
+ *
+ * @since 1.1
+ */
+public enum GcpEnvironment {
+	/**
+	 * Matches Kubernetes instances.
+	 */
+	KUBERNETES_ENGINE(() -> System.getenv("KUBERNETES_SERVICE_HOST") != null),
```

Comment:
this will move to provider; allows fallthrough; guarantees only one environment selected